### ISSUE TITLE
"That will overload the settings page."

### ISSUE THREAD

**Issue by dbauer-ets** [dbauer-ets/moodle-qtype_formulas#129]
For who? For the programmer or for the user?

If it's for the programmer, if there are programming-related objections, like difficulty managing the database and easier solutions exist, then I agree, depending on the case at hand.

If it's for the user, that argument doesn't seem justifiable. As you know, the Formulas question is primarily aimed at STEM teachers who are certainly capable of handling a page with numerous settings.

Moreover, almost all the settings in the Formulas question are optional. In fact, you can create a question with just three parameters: name, text, and answer. For example:
* Question name: 'question 1',
* Question text: 'What is 2 + 2?',
 * Answer: '4'

That's it. In this sense, the Formulas question is one of the easiest for beginners to tackle !

However, users progress quickly, and soon they are limited by the lack of options, which I temporarily supplement with snippets. In my opinion, we should add as many options as possible.

Oh no! Not as many options as possible! But why this unjustified fear? Obviously, it's more work. Is that a problem?

It gives the impression that every reason is good enough to avoid adding options when one doesn't want to. I think this is the new excuse for several programmers working on Moodle. They don't have the time or resources to carry out certain developments that would nevertheless be very useful. If that's the real reason, obviously it's a valid one. You can't do anything if you don't have the resources to do it.

Some programmers, however, hesitate to give this explanation and prefer to use the excuse of it being "a security matter". Since that excuse is ridiculous and worn out, the new excuse is that "it will overload the settings page", an equally ridiculous excuse.

In short, if we don't have the resources to do something, if there are good reasons not to do it, or if there's a better way to do it, then we don't do it.

However, if we have the resources and there are good reasons to do something, then we do it. Adding options to the Formulas question, as many options as possible, is a good reason. The goal is to make the Formulas question the best, most advanced, most flexible, and not the opposite.

Now that we have once again an excellent programmer, Philipp Imhof, who has taken over from Jean-Michel Védrine, I hope, for my part, to be able to instill the right direction in the development of the Formulas question.

**Comment by PhilippImhof** [comment]
> For who? For the programmer or for the user?

For both, obviously. And I have explained why.

> If it's for the user, that argument doesn't seem justifiable. As you know, the Formulas question is primarily aimed at STEM teachers who are certainly capable of handling a page with numerous settings.

People are capable of doing a lot of stuff. Just because they are able to use a cluttered form, that does not mean they like doing so. And it does not mean that it's good UX practice to clutter forms.

> Moreover, almost all the settings in the Formulas question are optional. In fact, you can create a question with just three parameters: name, text, and answer. 

Yes. And you have to scroll miles and miles to find those fields among the clutter one might add.

> However, users progress quickly, and soon they are limited by the lack of options, which I temporarily supplement with snippets. In my opinion, we should add as many options as possible.

I still do not get your point: no one, **absolutely no one** has ever said anything against the option to set the field width or the alignment (which you have brought up just recently).

> Oh no! Not as many options as possible! But why this unjustified fear? Obviously, it's more work. Is that a problem?

You are asking the guy who rewrites everything from scratch whether he is afraid of work? Seriously?

> It gives the impression that every reason is good enough to avoid adding options when one doesn't want to. 

I don't know how you come to that conclusion. I have suggested a clean and easy way to implement those options without cluttering the form. My approach also has the benefit of being rather easy to implement. These options **will** be added. I have already outlined how. And I have explained why that has to wait. I am not going to implement that in the legacy parser code only to redo it in a few months when the new version is ready. That would be a total waste of time and only slow down the much more important project currently going on.

> However, if we have the resources and there are good reasons to do something, then we do it. Adding options to the Formulas question, as many options as possible, is a good reason. The goal is to make the Formulas question the best, most advanced, most flexible, and not the opposite.

Well that's exactly what I am trying to do by investing countless hours into the development of this plugin. It has gone a long way in the last ~ 12 months and much more is yet to come.

**Comment by dbauer-ets** [comment]
> People are capable of doing a lot of stuff. Just because they are able to use a cluttered form, that does not mean they like doing so. And it does not mean that it's good UX practice to clutter forms.

A form that has many options is not necessarily "cluttered". It simply has many options. That is not a flaw, on the contrary.

> Yes. And you have to scroll miles and miles to find those fields among the clutter one might add.

"miles and miles", you're exaggerating way too much. A form can be long, but it is not miles and miles long. As I said above, almost all settings have a default value, so having many options is not a problem, it simply give more flexibility to the users. If the page gets too long, it is a simple matter to hide/show different sections.

> I still do not get your point: no one, absolutely no one has ever said anything against the option to set the field width or the alignment (which you have brought up just recently).

Yes, but maybe you should also be in favor of the many more options that can be included.

> You are asking the guy who rewrites everything from scratch whether he is afraid of work? Seriously?

I know very well that you work very hard. So the answer is no, you're absolutely not afraid of work. I might have expressed myself poorly. Instead of saying "Is that a problem?", I should have said something like "I know that's not a problem with you".

> I don't know how you come to that conclusion. I have suggested a clean and easy way to implement those options without cluttering the form. My approach also has the benefit of being rather easy to implement. These options will be added. I have already outlined how. And I have explained why that has to wait. I am not going to implement that in the legacy parser code only to redo it in a few months when the new version is ready. That would be a total waste of time and only slow down the much more important project currently going on.

Absolutely. I was not talking about any specific issue. You decide when to work on any specific issue.


I will try to reformulate my thought.

It seems to me that some programmers say "it'll clutter the form" a bit too often, and it's become some kind of trend. It's like whenever there's a question, the programmer's response is "it'll clutter the form". It's an argument that abruptly ends any conversation. "It'll clutter the form," so we can't do anything, implying I have other things that interest me more.

I'm not saying that you're the initiator of this trend, but perhaps a victim. I fear that you might have been influenced poorly by other programmers. I wish you could perhaps see things a bit differently.

Let me give you an example:

The core questions have received almost no development for years. They incorporate very few options. The "Calculated" question is messed up with incomprehensible menus spread across several pages. Furthermore, it's buggy without any programmer intervening. The reason given? It'll clutter the form or some other absurd reasons. And what does Moodle do? It turns to H5P! What a shame. Moodle can't develop its own questions !?

The Formulas question is a gem. And you're a rare gem of a programmer. I simply wish that you wouldn't be negatively influenced by other programmers who might be excellent at coding but have no vision regarding the development of questions.

I would never dare to criticize your exemplary, extraordinary work. I just wish to share with you my vision regarding the development of the question, and hopefully bring you to adhere to it. :-)

**Comment by PhilippImhof** [comment]
> A form that has many options is not necessarily "cluttered". It simply has many options. That is not a flaw, on the contrary.

I agree. It is not *necessarily* cluttered. But ours is (at least!) at the verge of being cluttered. Most of the fields we have in the edit form are necessary. They could not be removed without losing important functions that are being used on a regular basis. (Except maybe for "Other rules", which is probably not being used daily. That does not mean I want to remove it, though -- don't get me wrong.)

> "miles and miles", you're exaggerating way too much. A form can be long, but it is not miles and miles long. As I said above, almost all settings have a default value, so having many options is not a problem, it simply give more flexibility to the users. If the page gets too long, it is a simple matter to hide/show different sections.

Well of course it's not truly miles long. That was a metaphor. As I already said, the fields having default values does not help against the clutter. The field is still there and taking up space. Also note that one can hide entire sections (e.g. Part 1) and one can hide/show **one** part of a section by using the "Show less"/"Show more" link. 

My remark was aimed at your suggestion in #126 of having six more input fields per answer field. (Considering what you proposed in #45, it would make ~ 10 fields and they could not be on one row, so we would have 2-3 rows per answer field. For a feature that is absolutely useful in **certain** (maybe even many) situations, but also unnecessary in **many** other cases. That's what I meant: There is a way to offer simple access to that functionality for those who want it without worsening the UX for those who don't need it.

Also, I'd like to explicitly mention accessibility. Please do not forget that Moodle has a big focus on accessibility. We should not forget people who might use that form with a screenreader. (They are certainly a minority, but they deserve special consideration.)

> Yes, but maybe you should also be in favor of the many more options that can be included.

I am; see my answer in #45. But I'm not in favour of **all** the options you proposed.

> It seems to me that some programmers say "it'll clutter the form" a bit too often, and it's become some kind of trend. It's like whenever there's a question, the programmer's response is "it'll clutter the form". It's an argument that abruptly ends any conversation. "It'll clutter the form," so we can't do anything, implying I have other things that interest me more.

I see your point. I am not that kind of guy. I take feature requests seriously and analyse them. If I say something will clutter the form, then that's because I sincerely think it will. If I think that something is useless, I do say so rather than make up an excuse. And I am absolutely open to implementing stuff that I would personally not need. (In fact, I have done so many times already. Personally, I have absolutely no use for the whole unit things, because I teach pure mathematics. But I will rewrite it nevertheless, because I see that this is an important feature for many of our users.)

> I'm not saying that you're the initiator of this trend, but perhaps a victim. I fear that you might have been influenced poorly by other programmers. I wish you could perhaps see things a bit differently.

I see things differently quite often. But knowing that I am not an expert in every field, I am ready to listen to those who deal with certain problems regularly. If one often hears the same answer from different developers, there are two possible conclusions: either, they are all too stupid to understand my brilliant idea; or my idea is not as brilliant as I thought and those guys do really have a point.

> The core questions have received almost no development for years. They incorporate very few options. The "Calculated" question is messed up with incomprehensible menus spread across several pages. Furthermore, it's buggy without any programmer intervening. The reason given? It'll clutter the form or some other absurd reasons. And what does Moodle do? It turns to H5P! What a shame. Moodle can't develop its own questions !?

I do not really agree. It's true that the basic question types do not get much development. But sometimes they do; I, for example, have added functionality to allow HTML and MathJax in Moodle's "calculated multiple choice" question type and it's waiting for integration. 

If you feel that some question types lack options or are buggy, please open a report in the Moodle tracker. Or better yet, propose a patch to improve them. I have always had very good and constructive conversations with the folks from Moodle HQ.

Also, turning to H5P is not a shame. It's simply moving forward. Many schools want to have more modern and more interactive question types. H5P is an open standard. It would be absolutely stupid to reinvent the wheel. Integrating H5P into Moodle is a very good decision IMHO.

> The Formulas question is a gem. And you're a rare gem of a programmer. I simply wish that you wouldn't be negatively influenced by other programmers who might be excellent at coding but have no vision regarding the development of questions.

Thanks for the compliment. The Formulas question can indeed be a gem to many people. But bear in mind that is also awfully disgusting to many others, because there's quite a learning curve to start. In that sense, Formulas will never be a question type "for everyone".

I think that most people at Moodle HQ do absolutely have a vision. Looking at current Moodle and comparing it to versions from ~ 3-5 years ago shows that great progress has been made.

**Comment by PhilippImhof** [comment]
Closing this, as it is more a discussion than an open issue to work on.
